### PR TITLE
MSPSDS-1094 ts intro slides

### PR DESCRIPTION
**Pull request by MatthewBurstein**
## Checklist:

- [X] Automated checks are passing locally.
- [X] CHANGELOG updated if change is worth telling users about.
### General testing
- [X] Test without javascript
- [X] Test on small screen
### Accessibility testing
- [X] Works keyboard only
- [X] Tested with one screen reader
- [X] Zoom page to 400% - content still visible
- [ ] Disable css - does content make sense and appear in a logical order?
- [ ] Passes automated checker (automated in build or manual)

### CODE REVIEW

File: mspsds-web/app/controllers/introductions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class IntroductionsController < ApplicationController
+  include Wicked::Wizard
+
+  steps :overview, :report_products, :track_investigations, :share_data
+
+  before_action :set_has_viewed_introduction
```
Comment (broder) [inline mspsds-web/app/controllers/introductions_controller.rb]: Does this set that the user has viewed the introduction even if they've only seen the first step? Is that intentional?

If it's to do with the skip link, maybe we should make the skip link go to an action which sets the `has_viewed_introduction` flag and then redirects to root?

File: mspsds-web/app/controllers/introductions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class IntroductionsController < ApplicationController
+  include Wicked::Wizard
+
+  steps :overview, :report_products, :track_investigations, :share_data
```
Comment (broder) [inline mspsds-web/app/controllers/introductions_controller.rb]: I think this controller is almost simple enough that we don't need a wizard. Maybe we'd just need
```
def overview; end
def report_products; end
def track_investigations; end
def share_data
  User.current.has_viewed_introduction = true
end
```

Using a wizard is also fine though 🙂

File: mspsds-web/app/models/user.rb [hunk 3]
```diff
@@ -6,6 +6,12 @@ class User < Shared::Web::User
   has_many :team_users, dependent: :nullify
   has_many :teams, through: :team_users
 
+  has_one :user_attributes, dependent: :destroy
+
+  # Getters and setters for each UserAttributes column should be added here so they can be accessed directly
+  # from the User object via delegation.
+  delegate :has_viewed_introduction, :has_viewed_introduction=, to: :get_user_attributes
```
Comment (broder) [inline mspsds-web/app/models/user.rb]: Neat!

File: mspsds-web/app/models/user_attributes.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class UserAttributes < ApplicationRecord
+  belongs_to_active_hash :user
+
+  def has_viewed_introduction=(value)
```
Comment (broder) [inline mspsds-web/app/models/user_attributes.rb]: Will we ever want to set this to `false`? Maybe we should just define a method like
```
def has_viewed_introduction!
  update(has_viewed_introduction: true)
end
```

File: mspsds-web/app/views/introductions/overview.html.slim [hunk 5]
```diff
@@ -0,0 +1,15 @@
+- content_for :page_title, "Overview - Product safety database - GOV.UK"
```
Comment (broder) [inline mspsds-web/app/views/introductions/overview.html.slim]: What does the page title end up looking like? We're adding another bit [here](https://github.com/UKGovernmentBEIS/beis-mspsds/blob/master/mspsds-web/app/views/layouts/application.html.slim#L16) - does that result in "Overview - Product safety database - GOV.UK - Market Surveillance"?

File: mspsds-web/app/views/introductions/report_products.html.slim [hunk 6]
```diff
@@ -0,0 +1,12 @@
+- content_for :page_title, "Report products - Product safety database - GOV.UK"
+- content_for :after_header do
+  = render "skip_introduction"
+
+.flash-card
+  .govuk-grid-row
+    govuk-grid-column-two-thirds-from-desktop
```
Comment (broder) [inline mspsds-web/app/views/introductions/report_products.html.slim]: We're missing a `.` at the start of this classname

File: mspsds-web/app/views/introductions/overview.html.slim [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+- content_for :after_header do
+  = render "skip_introduction"
+
+.flash-card
```
Comment (broder) [inline mspsds-web/app/views/introductions/overview.html.slim]: Could / should we just apply this class to another one of the divs below or does it need to be on a separate div?

File: mspsds-web/config/routes.rb [hunk 8]
```diff
@@ -13,6 +13,8 @@
     end
   end
 
+  resources :introduction, controller: :introductions
```
Comment (broder) [inline mspsds-web/config/routes.rb]: Do we need to specify `only` here?

File: mspsds-web/test/system/keycloak_test.rb [hunk 9]
```diff
@@ -1,9 +1,18 @@
 require "application_system_test_case"
 
 class KeycloakTest < ApplicationSystemTestCase
+  setup do
+    sign_out_if_signed_in
```
Comment (broder) [inline mspsds-web/test/system/keycloak_test.rb]: Will we always be on a page when this is called? Should we call something like `Shared::Web::KeycloakClient.instance.logout` instead?

File: mspsds-web/app/controllers/introductions_controller.rb (hunk 1 above)
Comment (MatthewBurstein) [reply mspsds-web/app/controllers/introductions_controller.rb]: @edwardhorsford says the flag should be set as soon as they have seen any intro slide. This means that if they use the skip introduction link the first time then they won't be shown the slides the second time they log in.

File: mspsds-web/app/models/user_attributes.rb (hunk 4 above)
Comment (MatthewBurstein) [reply mspsds-web/app/models/user_attributes.rb]: Agreed this is nicer

File: mspsds-web/app/controllers/introductions_controller.rb (hunk 1 above)
Comment (broder) [reply mspsds-web/app/controllers/introductions_controller.rb]: We could use custom logic for the skip link though?

File: mspsds-web/app/controllers/introductions_controller.rb (hunk 2 above)
Comment (MatthewBurstein) [reply mspsds-web/app/controllers/introductions_controller.rb]: 👍

File: mspsds-web/config/routes.rb (hunk 8 above)
Comment (MatthewBurstein) [reply mspsds-web/config/routes.rb]: This turned out much better using `namespace` once I removed the wizard.

File: mspsds-web/app/views/introductions/overview.html.slim (hunk 7 above)
Comment (MatthewBurstein) [reply mspsds-web/app/views/introductions/overview.html.slim]: It needs to be a separate div. Otherwise the -15 margin on `.govuk-grid-row` is filled with the blue background of the flash card

File: mspsds-web/app/views/introductions/overview.html.slim (hunk 5 above)
Comment (MatthewBurstein) [reply mspsds-web/app/views/introductions/overview.html.slim]: Yes it does. @edwardhorsford confirmed the formatting from this PR so I've added it to application.html.slim instead.

File: mspsds-web/app/controllers/introductions_controller.rb (hunk 1 above)
Comment (MatthewBurstein) [reply mspsds-web/app/controllers/introductions_controller.rb]: Ah I see what you mean. Have done so now :)

File: mspsds-web/test/system/keycloak_test.rb (hunk 9 above)
Comment (MatthewBurstein) [reply mspsds-web/test/system/keycloak_test.rb]: I can't get this to work because direct calls to methods on `KeycloakClient` raise a `ProcCookieTokenNotDefined` error. 
I've not had a test run fail on `sign_out_if_signed_in` since adding it. Where do you think the problem might arise?

Comment (broder) [reply mspsds-web/test/system/keycloak_test.rb]: It might arise if this test is run first. Then you might not be on any page and therefore couldn't click the "Sign out" link. Maybe it's not something to worry about for now!

File: mspsds-web/test/controllers/homepage_controller_test.rb [hunk 10]
```diff
@@ -2,16 +2,24 @@
 
 class HomepageControllerTest < ActionDispatch::IntegrationTest
   setup do
-    mock_out_keycloak_and_notify
+    @user = mock_out_keycloak_and_notify
```
Comment (jasiekmiko) [inline mspsds-web/test/controllers/homepage_controller_test.rb]: I think I've stopped returning the user from this method in another branch, sorry! :S
Anything wrong with using `User.current` here?

File: mspsds-web/test/system/introduction_test.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
+require "application_system_test_case"
+
+class IntroductionTest < ApplicationSystemTestCase
+  setup do
+    @user = mock_out_keycloak_and_notify
```
Comment (jasiekmiko) [inline mspsds-web/test/system/introduction_test.rb]: ditto :)

File: mspsds-web/test/system/keycloak_test.rb [hunk 12]
```diff
@@ -1,9 +1,18 @@
...
+  end
+
+  teardown do
+    sign_out_if_signed_in
+  end
+
   test "can login" do
     visit root_path
     sign_in email: "admin@example.com", password: "password"
+    click_on "Skip introduction"
```
Comment (jasiekmiko) [inline mspsds-web/test/system/keycloak_test.rb]: This feels irrelevant to a "can login" test.
If it was added to make the test work, perhaps logging in as an opss user would make more sense?
If you are deliberately testing the "skip introduction" link works here, it could probably use its own test?

File: mspsds-web/app/assets/application/stylesheets/flashcard.scss [hunk 13]
```diff
@@ -0,0 +1,93 @@
+$black: #000000;
```
[the remark this answers is not in the document]
Comment (edwardhorsford) [reply mspsds-web/app/assets/application/stylesheets/flashcard.scss]: FWIW Black in GOV.UK services is never #000000 - so double reason to use the helpers.

File: mspsds-web/app/assets/application/stylesheets/flashcard.scss [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+.flash-card {
```
[the remark this answers is not in the document]
Comment (MatthewBurstein) [reply mspsds-web/app/assets/application/stylesheets/flashcard.scss]: They came from @edwardhorsford but I don't think we have a meaningful place to link to.

File: mspsds-web/app/assets/application/stylesheets/flashcard.scss (hunk 13 above)
[the remark this answers is not in the document]
Comment (MatthewBurstein) [reply mspsds-web/app/assets/application/stylesheets/flashcard.scss]: fixed :)

File: mspsds-web/app/views/introduction/track_investigations.html.slim [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+- content_for :after_header do
+  = render "skip_introduction"
+
+.flash-card
+  .govuk-grid-row
+    .govuk-grid-column-two-thirds-from-desktop
+      h1.govuk-heading-l Track and manage investigations
+      p.govuk-body-l
+        | Market surveillance authorities (MSAs) can report and manage routine product safety cases and track
+        |  their progress.
```
[the remark this answers is not in the document]
Comment (MatthewBurstein) [reply mspsds-web/app/views/introduction/track_investigations.html.slim]: 👍
